README.md
# WANetworkRouting

[![Version](https://img.shields.io/cocoapods/v/WANetworkRouting.svg?style=flat)](http://cocoapods.org/pods/WANetworkRouting)
[![License](https://img.shields.io/cocoapods/l/WANetworkRouting.svg?style=flat)](http://cocoapods.org/pods/WANetworkRouting)
[![Platform](https://img.shields.io/cocoapods/p/WANetworkRouting.svg?style=flat)](http://cocoapods.org/pods/WANetworkRouting)

**Developed and Maintained by [ipodishima](https://github.com/ipodishima) Founder & CTO at [Wasappli Inc](http://wasapp.li).**

**Sponsored by [Wisembly](http://wisembly.com/en/)**

A routing library to fetch objects from an API and map them to your app

- [x] Highly customizable: network request, authentication and mapping layers are separate, you can create your own.
- [x] Default network request layer built on top of AFNetworking 3.0
- [x] Default mapping layer built on top of WAMapping
- [x] Built-in object router
- [x] Built-in batch manager
- [x] Different configurations available
- [x] `NSProgress` support
- [x] Tested and used on real projects

Go visit the [wiki](https://github.com/wasappli/WANetworkRouting/wiki) for more details about `WANetworkRouting` advanced use.

`WANetworkRouting` is a library which turns `GET enterprises/:itemID` to `Enterprise : NSManagedObject` from a simple configuration step.

## Install and use
### Cocoapods
Use Cocoapods, this is the easiest way to install the router.

`pod 'WANetworkRouting'`

`#import <WANetworkRouting/WANetworkRouting.h>` 

### `WANetworkRoutingManager`
`WANetworkRoutingManager` is the core component of `WANetworkRouting`. It handles all the requests for you.

The initialization should contains at least a request manager. `WANetworkRouting` has a built in manager `WAAFNetworkingRequestManager` built on top of `AFNetworking 3.0`.

```objc
// Create a request manager
WAAFNetworkingRequestManager *requestManager = [WAAFNetworkingRequestManager new];

// Create the network routing manager 
WANetworkRoutingManager *routingManager = [WANetworkRoutingManager managerWithBaseURL:[NSURL URLWithString:@"http://baseURL.com"]
                                                                       requestManager:requestManager
                                                                       mappingManager:nil
                                                                authenticationManager:nil];

```

** Note ** This `routingManager` will have not mapping layer nor authentication layer. This example demonstrates how easy it is to test your API before configuring the mapping.

You can then perform any operations among `GET|POST|PUT|PATCH|DELETE|HEAD`. For eg:

- Fetch all the enterprises

```objc
[routingManager getObjectsAtPath:@"enterprises"
                      parameters:nil
                         success:^(WAObjectRequest *objectRequest, WAObjectResponse *response, NSArray *mappedObjects) {
                             NSDictionary *json = response.responseObject;
                             // Do something with the JSON
                         }
                         failure:^(WAObjectRequest *objectRequest, WAObjectResponse *response, id<WANRErrorProtocol> error) {
                         }];
```

- Update an enterprise

```objc
[routingManager putObject:nil
                     path:@"enterprises/1"
               parameters:@{@"key": @"newValue"}
                  success:^(WAObjectRequest *objectRequest, WAObjectResponse *response, NSArray *mappedObjects) {
                  }
                  failure:^(WAObjectRequest *objectRequest, WAObjectResponse *response, id<WANRErrorProtocol> error) {
                  }];
```

### Add mapping layer
This is where things become to be interesting

The idea is to add a mapping layer so that your json gets mapped as objects. This is done by using a default mapping service built on top of [`WAMapping`](https://github.com/wasappli/WAMapping). I strongly encourage you to read the docs about how to configure the mapping.

There are 3 steps:

- Configure the mapping,
- Configure the response descriptor,
- Optional: configure the request descriptor.

#### Step 1: configure the mapping

Have a look at [`WAMapping`](https://github.com/wasappli/WAMapping) for every details about the mapping:

```objc
// Specify a store to use between WAMemoryStore, WANSCodingStore, WACoreDataStore, your own store.
WAMemoryStore *memoryStore = [[WAMemoryStore alloc] init];
           
// Create the mapping description for `Enterprise`
WAEntityMapping *enterpriseMapping = [WAEntityMapping mappingForEntityName:@"Enterprise"];
enterpriseMapping.identificationAttribute = @"itemID";
[enterpriseMapping addAttributeMappingsFromDictionary:@{
                                                        @"id": @"itemID",
                                                        @"name": @"name",
                                                        @"address.street_number": @"streetNumber"}];
// Create the mapping manager
WAMappingManager *mappingManager = [WAMappingManager mappingManagerWithStore:memoryStore];
                                                                
WAAFNetworkingRequestManager *requestManager = [WAAFNetworkingRequestManager new];

// Create the network routing manager 
WANetworkRoutingManager *routingManager = [WANetworkRoutingManager managerWithBaseURL:[NSURL URLWithString:@"http://baseURL.com"]
                                                                       requestManager:requestManager
                                                                       mappingManager:mappingManager
                                                                authenticationManager:nil];
                                                                
// Add a default date formatter on mapper
id(^toDateMappingBlock)(id ) = ^id(id value) {
    if ([value isKindOfClass:[NSString class]]) {
        return [dateFormatter dateFromString:value];
    }

    return value;
};

[mappingManager addDefaultMappingBlock:toDateMappingBlock
                   forDestinationClass:[NSDate class]];
```

#### Step 2: configure the response descriptors
This will map a path with a mapping. For example, when fetching `/enterprises`, you would get a JSON as

```
{
	"array_of_enterprises": [{
		"name": "Enterprise 1"
	}, {
		"name": "Enterprise 2"
	}]
}
```

The request descriptor would then be as follows:

```objc
WAResponseDescriptor *enterprisesResponseDescriptor =
[WAResponseDescriptor responseDescriptorWithMapping:enterpriseMapping
                                             method:WAObjectRequestMethodGET
                                        pathPattern:@"enterprises" // The path on the URL
                                            keyPath:@"array_of_enterprises"]; // The key path to access the enterprises on JSON
```

For a `GET` or a `PUT` on an enterprise, which would be returned as

```
{
	"name": "Enterprise 1"
}
```

```objc
WAResponseDescriptor *singleEnterpriseResponseDescriptor =
[WAResponseDescriptor responseDescriptorWithMapping:enterpriseMapping
                                             method:WAObjectRequestMethodGET | WAObjectRequestMethodPUT // Specify multiple methods
                                        pathPattern:@"enterprises/:itemID" // The path
                                            keyPath:nil]; // The response would directly returns the object
```

Please note the syntax `enterprises/:itemID`: `:` means that the value would be replaced dynamically knowning that `itemID` is the property name on client side.

Finally, add the response descriptors to the mapping manager

```objc
[mappingManager addResponseDescriptor:enterprisesResponseDescriptor];
[mappingManager addResponseDescriptor:singleEnterpriseResponseDescriptor];
```

You can now use `WANetworkRouting` as this:

```objc
[routingManager getObjectsAtPath:@"enterprises"
                      parameters:nil
                         success:^(WAObjectRequest *objectRequest, WAObjectResponse *response, NSArray *mappedObjects) {
                             NSArray<Enterprise *> *enterprises = mappedObjects;
                             // Do something with the enterprises
                         }
                         failure:^(WAObjectRequest *objectRequest, WAObjectResponse *response, id<WANRErrorProtocol> error) {
                         }];
```

#### Step 3: configure the request descriptor
This step is optional, but allows you to reverse map an object to send it to the server.
Rather than creating by yourself the parameters dictionary on a `POST` for example, you could write:

```objc
WARequestDescriptor *enterpriseRequestDescriptor =
[WARequestDescriptor requestDescriptorWithMethod:WAObjectRequestMethodPOST
                                     pathPattern:@"enterprises" // The path on the URL
                                         mapping:enterpriseMapping
                                  shouldMapBlock:nil // On optional block which let you configure rather you want to reverse map a relation ship or not
                                  requestKeyPath:nil]; // The key path for the final dictionary
                                  
[mappingManager addRequestDescriptor:enterpriseRequestDescriptor];
```

Please note that the `requestKeyPath` fits with how your object would be wrapped on the parameters. If nil, the dictionary is the object as dictionary.

```
{
	"requestKeyPathValue": {
		"name": "Enterprise 1"
	}
}
```

This allows you to post an object like this

```objc
Enterprise *enterprise = [Enterprise new];
enterprise.name = @"Test";
        
[routingManager postObject:enterprise
                      path:@"enterprises"
                parameters:nil
                   success:^(WAObjectRequest *objectRequest, WAObjectResponse *response, NSArray *mappedObjects) {
                   }
                   failure:^(WAObjectRequest *objectRequest, WAObjectResponse *response, id<WANRErrorProtocol> error) {
                   }];
```

This will reverse map the `enterprise` to a dictionary and send it as parameters on `POST enterprise`. All magically without anymore work!

**Note**: If `enterprise` has no value for its `identificationAttribute` (see the mapping), the enterprise object would be automatically deleted from the store since the object returned by the server would be a new one with an `identificationAttribute`. Said in other words: you do not have to deal with potential duplicates.

### Add routing layer

We just saw this call

```objc
[routingManager postObject:enterprise
                      path:@"enterprises"
                parameters:nil
                   success:^(WAObjectRequest *objectRequest, WAObjectResponse *response, NSArray *mappedObjects) {
                   }
                   failure:^(WAObjectRequest *objectRequest, WAObjectResponse *response, id<WANRErrorProtocol> error) {
                   }];
```

Wouldn't it be great to write instead?

```objc
[routingManager postObject:enterprise
                      path:nil // No value for path
                parameters:nil
                   success:^(WAObjectRequest *objectRequest, WAObjectResponse *response, NSArray *mappedObjects) {
                   }
                   failure:^(WAObjectRequest *objectRequest, WAObjectResponse *response, id<WANRErrorProtocol> error) {
                   }];
```

Here comes the router!

```objc
WANetworkRoute *postEnterpriseRoute =
[WANetworkRoute routeWithObjectClass:[Enterprise class]
                         pathPattern:@"enterprises"
                              method:WAObjectRequestMethodPOST];

[routingManager.router postEnterpriseRoute];
```

Each time you are trying to `POST` an object of class `Enterprise`, it will supply the `enterprises` path for you.

Here is an other example with `GET` and `PUT` routes:

```objc
WANetworkRoute *enterpriseRoute =
[WANetworkRoute routeWithObjectClass:[Enterprise class]
                         pathPattern:@"enterprises/:itemID"
                              method:WAObjectRequestMethodGET | WAObjectRequestMethodPUT];

[routingManager.router addRoute:enterpriseRoute];
```

### Add authentication layer

This layer is important if your API has some authentication. Basically you should have a login/signup endpoint which returns some kind of token that you can renew if the requests gets an error like `401: token expired`.
By implementing the simple `WARequestAuthenticationManagerProtocol` protocol, and passing an instance of your class to the router manager, it will:

- Ask to authenticate the `NSMutableURLRequest`. Should be a token on `Authorization` HTTP header field.
- Ask if a request should be replayed: you received `401: token expired` for example then yes.
- Ask you to authenticate (renew the authorization somehow) and replay the request (`[routingManager enqueueRequest:]`) (The request will automatically be re authorized for you).

This will allow the routing manager to run every requests without any surprise when authentication has expired!

### Add batch 
The batch managers answers two needs:

1/ Let's say that you want to minimize the impact on server of having multiple calls to perform at the same time. Like `GET /me`, `GET /meetings`, `GET /configuration`
2/ Have offline support, for example add notes to a meeting and automatically sync them when you are back online!

Well, `WABatchManager` is there for you

#### Step 1: create a batch api
This library comes with a built-in batch manager from the following specifications. You can either create an API which conforms to the specification, or create a new batch manager from `WABatchManager` which fits your needs.

The specifications: (this is  closely following the batch api requests by facebook [Facebook batch API](https://developers.facebook.com/docs/graph-api/making-multiple-requests))

- Set a limit per session
- The data sent has the following format

```
{
	"batch": [{
		"uri": "\/meetings\/Ufd8f4e\/notes",
		"method": "POST",
		"body": "{\"note\":{\"hash\":\"817ebd76b6a89eff47bbd8e53f633c93eeadf23e\",\"title\":\"New first note\",\"type\":\"item\",\"position\":0}}",
		"headers": {
          "Content-Type": "application/json"
        }
	}, {
		"uri": "\/meetings\/Ufd8f4e\/notes",
		"method": "POST",
		"body": "{\"note\":{\"hash\":\"f4e893ab9a754c273707b264359e7ec12262959d\",\"title\":\"New second note\",\"type\":\"note\",\"position\":1}}"
	}]
}
```
Please note that `uri` is relative to the base URL and `body` is an encoded string

- Respond with

```
[
  {
    "body": "[{\"note\":\"...\"}]
    "code": 200
  },
  {
    "body": "[{\"note\":\"...\"}]
    "code": 200
  }
 ]
 ```

#### Step 2: create a batch manager

```objc
WABatchManager *batchManager = [[WABatchManager alloc] initWithBatchPath:@"/batch" limit:20];

routingManager = [WANetworkRoutingManager managerWithBaseURL:[NSURL URLWithString:kBaseURL]
                                              requestManager:requestManager
                                              mappingManager:mappingManager
                                       authenticationManager:authManager
                                                batchManager:batchManager];
```

#### Manually batch requests

```objc
// Create a batch session
WABatchSession *batchSession = [WABatchSession new];

// Enqueue the requests you want to perform
[batchSession addRequest:[WAObjectRequest requestWithHTTPMethod:WAObjectRequestMethodGET
                                                           path:@"/me"
                                                     parameters:nil
                                                optionalHeaders:nil]];
[batchSession addRequest:[WAObjectRequest requestWithHTTPMethod:WAObjectRequestMethodGET
                                                           path:@"/meetings"
                                                     parameters:nil
                                                optionalHeaders:nil]];
[batchSession addRequest:[WAObjectRequest requestWithHTTPMethod:WAObjectRequestMethodGET
                                                           path:@"/configuration"
                                                     parameters:nil
                                                optionalHeaders:nil]];
                                                
// Send the session
[batchManager sendBatchSession:batchSession
                  successBlock:^(id<WABatchManagerProtocol> batchManager, NSArray<WABatchResponse *> *batchResponses) {
                      
                  }
                  failureBlock:^(id<WABatchManagerProtocol> batchManager, WAObjectRequest *request, WAObjectResponse *response, id<WANRErrorProtocol> error) {
                      
                  }];
```

#### Add an offline mode
The offline mode works out of the box.
Basically, you register routes describing requests which can be enqueued if there is a network issue (you should not include `GET`) and you let the library do its job! It also integrates with mapping and authentication without any further step.

- Create routes to describe the requests to be enqueued

```objc
// Meetings
WANetworkRoute *modifyMeeting = [WANetworkRoute routeWithObjectClass:nil
                                                         pathPattern:@"meetings/:itemID"
                                                              method:WAObjectRequestMethodPUT | WAObjectRequestMethodeDELETE];

// Action items
WANetworkRoute *postActionItem = [WANetworkRoute routeWithObjectClass:nil
                                                          pathPattern:@"meetings/:itemID/notes"
                                                               method:WAObjectRequestMethodPOST];
```
- Add them to the batch manager

```objc
[batchManager addRouteToBatchIfOffline:modifyMeeting];
[batchManager addRouteToBatchIfOffline:postActionItem];
```

- Answers to errors

```objc
[self.apiManager putObject:meeting
                      path:nil
                parameters:nil
                   success:...
                   failure:^(WAObjectRequest *objectRequest, WAObjectResponse *response, id<WANRErrorProtocol> error) {
                       SLDMeeting *meetingReturned = nil;
                       // Check if the error is a batch one
                       if ([[error.originalError.domain isEqualToString:WANetworkRoutingManagerErrorDomain] && error.originalError.code == WANetworkRoutingManagerErrorRequestBatched) {
                           // Return the meeting for any UI update
                           meetingReturned = meeting;
                           // Store a "has offline changes" flag
                           meeting.hasOfflineModifications = @YES;
                           // Save the store
                           [store save];
                       }
                       
                       if (completion) {
                           completion(meetingReturned, error);
                       }
                   }];
```

- Wait for flush. The library automatically flush the offline changes on any request made while online. You can also manually trigger the flush with 

```objc
if ([batchManager needsFlushing]) {
  [batchManager flushDataWithCompletion:completion];
}
```

- Perform actions post flush

```objc
[batchManager setOfflineFlushSuccessBlock:^(id <WABatchManagerProtocol> batchManager, NSArray<WABatchResponse *> *batchResponses) {
    if (![batchManager needsFlushing]) {
        // Consider all meetings synchronized. You could also iterate trough batch responses
        NSArray *notSyncedMeetings = [SLDMeeting findAll];
        for (SLDMeeting *meeting in notSyncedMeetings) {
            meeting.hasOfflineModifications = @NO;
        }
        
        [store save];
    }
}];
```

## Progress

You can track the progress of your request. This library uses `NSProgress` class which is a great tool for dealing with progress.
If your app supports iOS 9+, there is the explicit child feature. You can then write:

(Please note that we are using a subclass of `NSProgress` that you can find on the sample. This is because `NSProgress` does not allow you to check if a progress already has a child :/)

```objc
WAProgress *mainProgress = [[WAProgress alloc] initWithParent:nil userInfo:nil];

CHANGELOG.md
## 1.0.0 - Released Jul 27, 2016
- Fixed an issue with response descriptors and keypaths… Wasn't working!

## 0.0.9 - Released Jun 10, 2016
- Send the WAObjectRequest for authentication

## 0.0.8 - Released Apr 27, 2016
- Added support for watch os

## 0.0.7 - Released Apr 22, 2016
### Pods
- Updated `AFNetworking` from `3.0` to `3.1`
- Updated `WAMapping` from `0.0.6` to `0.0.7`

### Batch Manager
- Added `canEnqueue` behavior from response and error
- Enqueue on `Offline` or `Time out` error on `NSURLErrorDomain`
- Enqueue on http status code of `500` (server down) or `503` (maintenance) 

## 0.0.6 - Released Mar 25, 2016
- Added Batch manager
- Fixed an issue on route and `DELETE`: the object is no longer mapped as a request
- Fixed an issue on request descriptors and routes with keypath. You can now write `meetings/:meeting.itemID/privatenotes/:itemID`
- Added reachability on `WARequestManagerProtocol` and `WAAFNetworkingRequestManager`: `isReachable`
- When you delete an object without using routes|request descriptors, the path is parsed to get the object ID to delete. It was not deleted before
- Fixed crashes when object requests have no success or failure blocks
- Fixed a crash when your request descriptor have no `shouldMapBlock`
 
## 0.0.5 - Released Mar 25, 2016
Fixed an issue with mapping progress when the object to map is empty

## 0.0.4 - Released Mar 22, 2016
Fixed an issue with mapping progress with several response descriptors

## 0.0.3 - Released Mar 21, 2016

This release brings the progress for download and mapping using `NSProgress`!

```objc
[routingManager getObjectsAtPath:@"posts"
                      parameters:nil
                        progress:^(WAObjectRequest *objectRequest, NSProgress *uploadProgress, NSProgress *downloadProgress, NSProgress *mappingProgress) {
                            NSLog(@"downloadProgress: %f -> %@ — mappingProgress: %f -> %@", downloadProgress.fractionCompleted, downloadProgress.localizedDescription, mappingProgress.fractionCompleted, mappingProgress.localizedDescription);
                        }
                         success:^(WAObjectRequest *objectRequest, WAObjectResponse *response, NSArray *mappedObjects) {
                         }
                         failure:^(WAObjectRequest *objectRequest, WAObjectResponse *response, id<WANRErrorProtocol> error) {
                         }];
```

By adding support of `NSProgress`, you can easily cancel the request. You have to call `[downloadProgress cancel]` and/or `[mappingProgress cancel]` depending on where you are on the task.

- Added a simple sample
- You no longer have to create `WAMapper` or `WAReverseMapper`. Simply pass the store.

Was

```objc
WAMemoryStore *memoryStore = [[WAMemoryStore alloc] init];

// Create both a mapper and a reverse mapper
WAMapper *mapper               = [WAMapper newMapperWithStore:memoryStore];
WAReverseMapper *reverseMapper = [WAReverseMapper new];

// Create the mapping manager
WAMappingManager *mappingManager = [WAMappingManager mappingManagerWithMapper:mapper
                                                                reverseMapper:reverseMapper];
```
is

```objc
WAMemoryStore *memoryStore = [[WAMemoryStore alloc] init];

// Create the mapping manager
WAMappingManager *mappingManager = [WAMappingManager mappingManagerWithStore:memoryStore];
```

- For registering default mapping blocks. 

Was

```objc
// Add a default date formatter on mapper
id(^toDateMappingBlock)(id ) = ^id(id value) {
    if ([value isKindOfClass:[NSString class]]) {
        return [dateFormatter dateFromString:value];
    }

    return value;
};

[mapper addDefaultMappingBlock:toDateMappingBlock
           forDestinationClass:[NSDate class]];
```
is

```objc
// Add a default date formatter on mapper
id(^toDateMappingBlock)(id ) = ^id(id value) {
    if ([value isKindOfClass:[NSString class]]) {
        return [dateFormatter dateFromString:value];
    }

    return value;
};

[mappingManager addDefaultMappingBlock:toDateMappingBlock
                   forDestinationClass:[NSDate class]];
```


## 0.0.2 - Released Feb 29, 2016
- Fixed an import with missing .h on main header
- Do not ask the store to delete an object if `nil`

## 0.0.1 - Released Feb 26, 2016
Initial release

Files/WABatchManager.m
//
//  WABatchManager.m
//  WANetworkRouting
//
//  Created by Marian Paul on 29/03/2016.
//  Copyright © 2016 Wasappli. All rights reserved.
//

#import "WABatchManager.h"

#import "WABatchCache.h"
#import "WABatchObject.h"
#import "WABatchSession.h"
#import "WAObjectRequest.h"
#import "WAObjectResponse.h"
#import "WAURLResponse.h"
#import "WABatchResponse.h"
#import "WANRErrorProtocol.h"

#import "WANetworkRoute.h"
#import "WANetworkRoutePattern.h"

#import "WANetworkRoutingMacros.h"

@interface WABatchManager ()

@property (nonatomic, strong) WABatchCache *cache;
@property (nonatomic, strong) NSMutableSet<WANetworkRoute *> *offlineRoutes;
@property (nonatomic, strong) NSString *batchPath;
@property (nonatomic, assign) NSUInteger batchLimit;

@property (nonatomic, assign, getter=isFlushing) BOOL flushing;

@end

@implementation WABatchManager
@synthesize delegate = _delegate;
@synthesize offlineFlushFailureBlock = _offlineFlushFailureBlock;
@synthesize offlineFlushSuccessBlock = _offlineFlushSuccessBlock;

- (instancetype)initWithBatchPath:(NSString *)batchPath limit:(NSUInteger)limit {
    WANRClassParameterAssert(batchPath, NSString);
    WANRParameterAssert(limit > 0);
    
    self = [super init];
    
    if (self) {
        self->_cache         = [[WABatchCache alloc] init];
        self->_offlineRoutes = [NSMutableSet set];
        self->_batchPath     = batchPath;
        self->_batchLimit    = limit;
    }
    
    return self;
}

+ (instancetype)batchManagerWithBatchPath:(NSString *)batchPath limit:(NSUInteger)limit {
    return [[self alloc] initWithBatchPath:batchPath limit:limit];
}

#pragma mark - WABatchManagerProtocol

#pragma mark Offline

- (BOOL)needsFlushing {
    return
    (([[self.cache batchObjects] count] == 0) ? NO : YES)
    &&
    !self.isFlushing;
}

- (void)flushDataWithCompletion:(void (^)(BOOL success))completion {
    
    if (![self needsFlushing]) {
        if (completion) {
            completion(YES);
        }
        return;
    }
    
    NSArray <WABatchObject *> *batchObjects = [self.cache batchObjects];
    if ([batchObjects count] > self.batchLimit) {
        batchObjects = [batchObjects subarrayWithRange:NSMakeRange(0, self.batchLimit)];
    }
    
    WABatchSession *batchSession = [WABatchSession new];
    for (WABatchObject *batchObject in batchObjects) {
        [batchSession addBatchObject:batchObject];
    }
    
    self.flushing = YES;
    
    wanrWeakify(self);
    [self sendBatchSession:batchSession
              successBlock:^(id <WABatchManagerProtocol> batchManager, NSArray<WABatchResponse *> *batchResponses) {
                  wanrStrongify(self);
                  self.flushing = NO;
                  
                  // Remove the batch objects
                  [self.cache removeBatchObjects:batchObjects];
                  
                  // Tell something that we sent a batch session
                  if (self.offlineFlushSuccessBlock) {
                      self.offlineFlushSuccessBlock(batchManager, batchResponses);
                  }
                  
                  // If we still needs flushing, continue
                  if ([self needsFlushing]) {
                      [self flushDataWithCompletion:completion];
                  }
                  else {
                      // We, the flush is a success!
                      if (completion) {
                          completion(YES);
                      }
                  }
              } failureBlock:^(id <WABatchManagerProtocol> batchManager, WAObjectRequest *request, WAObjectResponse *response, id<WANRErrorProtocol> error) {
                  wanrStrongify(self);
                  self.flushing = NO;
                  
                  if (self.offlineFlushFailureBlock) {
                      self.offlineFlushFailureBlock(batchManager, request, response, error);
                  }
                  
                  if (completion) {
                      completion(NO);
                  }
              }];
}

- (BOOL)canEnqueueOfflineRequest:(WAObjectRequest *)request withResponse:(WAObjectResponse *)response error:(id<WANRErrorProtocol>)error {
    __block BOOL matchesRoute = NO;
    
    [self.offlineRoutes enumerateObjectsUsingBlock:^(WANetworkRoute *obj, BOOL * _Nonnull stop) {
        BOOL correctMethod = obj.method & request.method;
        if (correctMethod) {
            BOOL samePath = [obj.pathPattern isEqualToString:request.path];
            if (!samePath && [obj.pathPattern rangeOfString:WANetworkRoutePatternObjectPrefix].location != NSNotFound) {
                WANetworkRoutePattern *pattern = [[WANetworkRoutePattern alloc] initWithPattern:obj.pathPattern];
                samePath = [pattern matchesRoute:request.path];
            }
            
            if (samePath) {
                matchesRoute = YES;
                *stop   = YES;
            }
        }
    }];
    
    BOOL matchesError = NO;
    
    if (!error && !response) {
        matchesError = YES;
    } else {
        // If the server is not responding or in maintenance mode, then enqueue
        if (response.urlResponse.statusCode == 500 || response.urlResponse.statusCode == 503) {
            matchesError = YES;
        }
        
        if ([[[error originalError] domain] isEqualToString:NSURLErrorDomain]) {
            NSInteger code = [[error originalError] code];
            if (code == NSURLErrorNotConnectedToInternet || code == NSURLErrorTimedOut) {
                matchesError = YES;
            }
        }
    }
    
    return matchesRoute && matchesError;
}

- (void)enqueueOfflineRequest:(WAObjectRequest *)request {
    NSAssert([self canEnqueueOfflineRequest:request withResponse:nil error:nil], @"Cannot enqueue %@ for offline batch, define a route to match the pattern", request.path);
    
    [self.cache addBatchObject:[self batchObjectFromRequest:request]];
}

- (void)reset {
    [self.cache dropDatabase];
}

#pragma mark Batch session

- (void)sendBatchSession:(WABatchSession *)session successBlock:(WABatchManagerSuccess)successBlock failureBlock:(WABatchManagerFailure)failureBlock {
    NSAssert(self.delegate, @"Delegate should no be nil, batch request needs to be handled");
    
    // Build the parameters from the session
    NSMutableArray *requestsAsDics = [NSMutableArray array];
    
    for (WABatchObject *batchObject in session.batchObjects) {
        NSMutableDictionary *requestAsDic = [NSMutableDictionary dictionary];
        requestAsDic[@"method"]           = batchObject.method;
        requestAsDic[@"uri"]              = [batchObject.uri hasPrefix:@"/"] ? batchObject.uri : [@"/" stringByAppendingString:batchObject.uri];
        
        if (batchObject.parameters) {
            // Encode it as string
            NSData *dJSON = [NSJSONSerialization dataWithJSONObject:batchObject.parameters
                                                            options:0
                                                              error:nil];
            NSString *json = [[NSString alloc] initWithData:dJSON encoding:NSUTF8StringEncoding];
            requestAsDic[@"body"] = json;
        }
        
        NSMutableDictionary *headers = [NSMutableDictionary dictionaryWithDictionary:batchObject.headers];
        headers[@"Content-Type"] = @"application/json";
        requestAsDic[@"headers"] = headers;
        
        [requestsAsDics addObject:requestAsDic];
    }
    
    WAObjectRequest *objectRequest = [WAObjectRequest requestWithHTTPMethod:WAObjectRequestMethodPOST
                                                                       path:self.batchPath
                                                                 parameters:@{@"batch": requestsAsDics}
                                                            optionalHeaders:nil];
    
    wanrWeakify(self);
    [objectRequest setCompletionBlocksWithSuccess:
     ^(WAObjectRequest *objectRequest, WAObjectResponse *fromBatchResponse, NSArray *mappedObjects) {
         wanrStrongify(self);
         
         NSArray *responses = (NSArray *)fromBatchResponse.responseObject;
         NSMutableArray *batchResponses = [NSMutableArray array];
         
         // Go through all responses
         NSUInteger index = 0;
         
         for (NSDictionary *dic in responses) {
             WAObjectResponse *response = [WAObjectResponse new];
             
             NSString *body = dic[@"body"];
             if ([body isEqual:[NSNull null]]) {
                 body = nil;
             }
             
             if (body) {
                 response.responseObject = [NSJSONSerialization JSONObjectWithData:[body dataUsingEncoding:NSUTF8StringEncoding]
                                                                           options:0
                                                                             error:nil];
             }
             
             WAURLResponse *urlResponse   = [WAURLResponse new];
             urlResponse.statusCode       = [dic[@"code"] integerValue];
             urlResponse.httpHeaderFields = dic[@"headers"];
             response.urlResponse         = urlResponse;
             
             WABatchObject *batchObject = session.batchObjects[index];
             WAObjectRequest *request   = [self objectRequestFromBatchObject:batchObject];
             
             WABatchResponse *batchResponse = [WABatchResponse new];
             batchResponse.response         = response;
             batchResponse.request          = request;
             
             [batchResponses addObject:batchResponse];
             index++;
         }
         
         [self.delegate batchManager:self haveBatchResponsesToProcess:[batchResponses copy]];
         
         if (successBlock) {
             successBlock(self, [batchResponses copy]);
         }
     }
                                          failure:
     ^(WAObjectRequest *objectRequest, WAObjectResponse *response, id<WANRErrorProtocol> error) {
         if (failureBlock) {
             failureBlock(self, objectRequest, response, error);
         }
     }];
    
    [self.delegate batchManager:self haveBatchRequestToEnqueue:objectRequest];
}
#pragma mark - Helpers

- (WABatchObject *)batchObjectFromRequest:(WAObjectRequest *)request {
    WABatchObject *batchObject = [WABatchObject new];
    batchObject.parameters     = request.parameters;
    batchObject.headers        = request.headers;
    batchObject.uri            = request.path;
    batchObject.method         = WAStringFromObjectRequestMethod(request.method);
    
    return batchObject;
}

- (WAObjectRequest *)objectRequestFromBatchObject:(WABatchObject *)batchObject {
    WAObjectRequest *request = [WAObjectRequest requestWithHTTPMethod:WAObjectRequestMethodFromString(batchObject.method)
                                                                 path:batchObject.uri
                                                           parameters:batchObject.parameters
                                                      optionalHeaders:batchObject.parameters];
    return request;
}

#pragma mark - Public methods

- (void)addRouteToBatchIfOffline:(WANetworkRoute *)route {
    WANRClassParameterAssert(route, WANetworkRoute);
    
    NSAssert(![self.offlineRoutes containsObject:route], @"You cannot add twice the same route");
    
    [self.offlineRoutes addObject:route];
}

@end
